graphistry/compute/ai/symbolic.py

```diff
@@ -0,0 +1,881 @@
+    def __call__(self, wrp_self, wrp_params, *args, **kwds):
+        super().override_reserved_signature_keys(wrp_params, *args, **kwds)
+        return "| {} SPL:".format(str(wrp_self))
+
+
+NEXT = " =>"
+
+
+class SplunkPrompts(ai.Prompt):
+    # https://docs.splunk.com/Documentation/SCS/current/SearchReference/SearchCommandExamples
+    def __init__(self) -> ai.Prompt:
+        super().__init__(
+            [
+                "search the main index where sourcetype is access_combined limit 10"
+                + NEXT
+                + "search index=main sourcetype=access_combined | head 10",
+                "Modify the following Splunk query that is useful for identifying malware in FireEye logs to instead use Palo Alto Networks logs:"
+                + NEXT
+                + "index=palo_alto sourcetype=pan:logs (action=allow OR action=deny) | eval src_ip=src_ip | eval dest_ip=dest_ip | where (src_ip=* OR dest_ip=*) | stats count by src_ip, dest_ip, action | sort - count",
+                "Lets try to find out how many errors have occurred on the Buttercup Games website"
+                + NEXT
+                + "buttercupgames (error OR fail* OR severe)",
+                "write a splunk query for the index `redteam_50k` that uses the src and dst information to output a table for events where RED=1. you can use closest matching fields from [src_computer, other, dst_computer, time]"
+                + NEXT
+                + '| search index="redteam_50k" RED=1 | Table src_computer, dst_computer',
+            ]
+        )
+
+
+class Splunk(ai.Expression):
+    @property
+    def static_context(self):
+        return SPLUNK_CONTEXT
+
+    def forward(self, sym: ai.Symbol, *args, **kwargs):
+        @ai.few_shot(
+            prompt="Generate queries based on the Splunk SPL domain specific language description\n",
+            examples=SplunkPrompts(),
+            pre_processor=[SplunkPreProcessor()],
+            post_processor=[StripPostProcessor()],
+            # stop=[''],
+            **kwargs,
+        )
+        def _func(_) -> str:
+            pass
+        return self._sym_return_type(_func(Splunk(sym)))
+
+    @property
+    def _sym_return_type(self):
+        return Splunk
+
+    def as_splunk(self):
+        return self.value
+
+
+class AIGraph(ai.Expression):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mem = {}
+        # self.process = self._process
+
+    def process(self, data, *args, **kwargs):
+        return process(data, *args, **kwargs)
+
+    def cache_sym(self, *args, **kwargs):
+        key = args[0]
+        value = args[1]
+        self.mem[key] = value
+
+    # @cache_sym
+    def _get_likely_src_dst(self, df):
+        srcdst = self.query(
+            f"what are the likely src and dst columns of this dataframe? {df.columns}\n return as [src, dst]"
+        ).list("[]", constraint=lambda x: len(x) == 2)
+        return srcdst
```

Comment:
maybe there phrasing is, "which columns represent entities to display as nodes?"

if 2, then src<>dst

if > 2, then hypergraph

if < 2, then ???